src/librustc/middle/traits/error_reporting.rs

```diff
@@ -62,6 +66,73 @@ pub fn report_projection_error<'a, 'tcx>(infcx: &InferCtxt<'a, 'tcx>,
     }
 }
 
+fn report_on_unimplemented<'a, 'tcx>(infcx: &InferCtxt<'a, 'tcx>,
+                                     trait_ref: &TraitRef<'tcx>) -> Option<String> {
+    let def_id = trait_ref.def_id;
+    let mut report = None;
+    ty::each_attr(infcx.tcx, def_id, |item| {
+        if item.check_name("on_unimplemented") {
+            match item.meta().node {
+                ast::MetaItem_::MetaNameValue(_, ref lit) => match lit.node {
+                    ast::Lit_::LitStr(ref i, _) => {
+                        let def = ty::lookup_trait_def(infcx.tcx, def_id);
+                        let mut generic_map = def.generics.types.iter_enumerated()
+                                                 .map(|(param, i, gen)| {
+                                                       (gen.name.as_str().to_string(),
+                                                        trait_ref.substs.types.get(param, i)
+                                                                 .user_string(infcx.tcx))
+                                                      }).collect::<HashMap<String, String>>();
+                        generic_map.insert("Self".to_string(),
+                                           trait_ref.self_ty().user_string(infcx.tcx));
+                        let parser = Parser::new(i.get());
+                        let mut errored = false;
+                        let err: String = parser.filter_map(|p| {
+                            match p {
+                                Piece::String(s) => Some(s),
+                                Piece::NextArgument(a) => match a.position {
+                                    Position::ArgumentNamed(s) => match generic_map.get(s) {
+                                        Some(val) => Some(val.as_slice()),
+                                        None => {
+                                            infcx.tcx.sess
+                                                 .span_err(item.meta().span,
+                                                           format!("There is no type parameter \
```

Comment:
Error messages conventionally start with a lower-case letter.

(They're printed like `error: There is no type ...`, and that looks a bit peculiar.)